d2/src/main/java/com/linkedin/d2/balancer/servers/ZooKeeperAnnouncer.java

```diff
@@ -268,6 +268,7 @@ public synchronized void markUp(final Callback<None> callback)
 
   private synchronized void doMarkUp(Callback<None> callback)
   {
+    _indisAnnouncer.emitAnnouncement(_cluster, _uri.getHost(), _uri.getPort(), IndisAnnouncer.HealthStatus.READY);
```

Comment:
Here you don't have the uri properties yet so the announcement can't happen. It should happen in the ephemeralStore class where all the complex logic about checks, warmups, building uriProperties, etc, have been done (in the ZooKeeperAnnouncer and ZooKeeperServer).
(Here the markUp could happen to warmupCluster too, please take a look at the warmup-related code in this class. So using _cluster here is not right. If you do the announcement in the ephemeralStore class, the exact cluster/prop name will be passed into it, so you don't need to worry about it.)

Reply:
Makes sense, I have pulled the IndisAnnouncer down into ZookeeperServer which, as you say, actually has all the info we need to make an announcement. Thanks for the pointer!

Reply:
What is the feasibility of implementing an alternative server to `ZooKeeperServer`? It doesn't feel right to have the `IndisAnnouncer`, which is itself an abstraction over any underlying technology, to be used from a ZooKeeper-specific concrete implementation of `LoadBalancerServer`.

Reply:
we're just piggy-backing zookeeper logic for **triggering the announcement** for now just to keep the announcement behaviors the same as ZK in the transition phase (rolling out INDIS on top of ZK d2 data). After the transition phase, we'll deprecate the whole ZookeeperAnnouncer -- ZookeeperServer -- ZKStore system and build a INDIS readiness manager in container which manages the data and timing of the announcement and integrates with INDISAnnouncer.